# SOTA Spots Map
by www.operator-paramedyk.pl / SQ9NIL

explained step-by-step

## What SOTA is?
SOTA (Summits On The Air) is an activity designed for radio amateurs (called also HAMs) who like hiking. It is about communication via ration between an activator - operator who climbed a designated summit (map available at https://sotl.as/map) - and chasers - all other operators. Call (or QSO) may be done with the use of telegraphy (morse code), voice or data transmission on any band available for radio amateurs. To make a successfull call, both operators need to exchange report, which say how they hear each other, and log it into a log. SOTA results are then uploaded into https://www.sotadata.org.uk/en/ webpage.

If you hear someone activating a summit (or want to make chasers aware the you are currently activating), you can send a spot for dedicated webpage (https://sotawatch.sota.org.uk/en/).

You can find all information about SOTA programme at https://www.sota.org.uk/.

This script is designed for chasers who want to check which summits are currently activated. With the use of data of latest activations received via SOTA API and summits coordinates in SOTA Database file, it creates an interactive map to visualise frequency and band of each activation.

For best experience, you can run this script during weekends, when there's a spike in SOTA activity.

You may also find helpful live version of this script I deployed at my webpage: https://www.operator-paramedyk.pl/sota/.

## Script algorithm

### Import of relevant packages
There are seven packages to be installed for this script, all specified in requirements.txt file. You can install all of them at once typing in terminal following line:

```pip install -r requirements.txt```

If you use Google Colab (https://colab.research.google.com/), all of them should be installed by default, so you can import them directly.

In [58]:
import requests # for communication with API
import pandas as pd # for data analysis
from datetime import datetime, timedelta # for time calculations
import folium # for data visualisation on a map

### Bands and modes
Before the analysis of SOTA data, I'll create two dataframes to store bands and modes available for HAM radio amateurs. Less popular ones - below 1.8 MHz (wavelength > 180 m) and above 900 MHz (wavelength < 30 cm) will be clustered together. To each band and mode I'll assign a color to be used for further visualisation.

In [59]:
# create dataframes to store bands assign them to colors for visualisation
# lower and upper freqs does not refer exactly to bandplan to make sure frequencies are mapped correctly during visualisation
bands = {
    'band': ['1.8 MHz or below', '3.5 MHz', '5 MHz', '7 MHz', '10 MHz', '14 MHz', '18 MHz', '21 MHz', '24 MHz', '28 MHz', '50 MHz', '70 MHz', '144 MHz', '220 MHz', '433 MHz', '900 MHz or above'],
    'lower_freq': [0, 3, 4.5, 6, 9, 13, 16, 19, 24, 27, 45, 65, 142, 210, 420, 850],
    'upper_freq': [2.5, 4, 5.5, 8, 11, 15, 18.5, 23, 26, 35, 55, 75, 148, 240, 460, 500000],
    'color': ['saddlebrown','chocolate', 'brown','red', 'salmon', 'orange', 'darkkhaki', 'yellow', 'olivedrab', 'green', 'lime', 'cyan', 'blue', 'purple', 'magenta', 'pink'],
}
bands_df = pd.DataFrame(bands)

bands_df = bands_df.set_index('band')
bands_df['color'] = bands_df['color'].astype('string')

In [60]:
# create dataframes to store modes and assign them to colors for visualisation
modes = {
    'mode': ['AM', 'CW', 'Data', 'DV', 'FM', 'SSB', 'Other'],
    'color': ['lime', 'red', 'cyan', 'magenta', 'yellow', 'blue', 'orange']
}

modes_df = pd.DataFrame(modes)
modes_df['color'] = modes_df['color'].astype('string')
modes_df['mode'] = modes_df['mode'].astype('string')

### Get activations data

Now I'll create a function to download data on latest activations with the use of SOTA API. Argument used during function execution is used in API request to define spots in scope. If the number is **negative**, it defines the timerange (in hours) of spots to be downloaded (by default function looks for the spots sent in latest 1 hour). If it's **positive**, the asked number of latest spots will be provided.

If you use negative number and no spots will be found, function will download latest 10 spots to make sure there are data to be analysed.

Spots downloaded are converted via JSON format to a dictionary, which is returned by the function.

In [61]:
def get_spots(time = -1):
      """Downdload SOTA spots sent in defined timeframe or defined number of latests spots and returns them as dictionary"""
      # if time is negative - download spots alerted in defined number of alerts
      # if time is positive - download given number of latest spots
      # by default looking from a spots sent in last 1 hour
      temp_spots_dict = {}
      url = f'https://api2.sota.org.uk/api/spots/{time}/all'
      r = requests.get(url)
      print(f'Status code: {r.status_code}')
      temp_spots_dict = r.json()
      if time > 0:
            print(f'{len(temp_spots_dict)} found where expected number was {time}.')
      if time <= 0:
            print(f'{len(temp_spots_dict)} spots found in latest {-time} h.')
      # if there are no spots sent in time provided, return latest 10 to make sure dictionary is not empty
      if len(temp_spots_dict) == 0:
          temp_spots_dict = get_spots(10)
      return temp_spots_dict


Now we'll use this function to download the spots sent in last hour and convert the dictionary into DataFrame.

In [62]:
# import spots and convert them into DataFrame
spots_dict = get_spots(-1)

spots_df = pd.DataFrame(spots_dict)

Status code: 200
17 spots found in latest 1 h.


Single spot looks good:

In [63]:
spots_dict[0]

{'id': 80877,
 'userID': 19344,
 'timeStamp': '2025-04-14T21:42:02',
 'comments': '[RBNHole] at VE6JY 19 WPM 22 dB SNR',
 'callsign': 'RBNHOLE',
 'associationCode': 'W7O',
 'summitCode': 'WV-144',
 'activatorCallsign': 'N7KOM',
 'activatorName': 'Tim',
 'frequency': '7.062',
 'mode': 'CW',
 'summitDetails': 'Lenhart Butte, 645m, 1 points',
 'highlightColor': None}

but if you look at the datatypes in dataframe, all information is saved as object type.

In [64]:
spots_df.dtypes

id                    int64
userID                int64
timeStamp            object
comments             object
callsign             object
associationCode      object
summitCode           object
activatorCallsign    object
activatorName        object
frequency            object
mode                 object
summitDetails        object
highlightColor       object
dtype: object

To make furtther analysis possible, we need to convert the data to string, numeric or date format. Also, I'll create one more column to get full summit reference (country/region-summit ID), which we'll use later to match between spot and SOTA database.

After this, we can confirm all data in spots_df DataFrame has appropriate type.

In [67]:
# convert frequency errors to 0
spots_df.loc[~spots_df['frequency'].str.match(r'\d+(\.\d+)?'), 'frequency'] = 0

# convert datatypes for relevant fields
spots_df['activatorCallsign'] = spots_df['activatorCallsign'].astype('string')
spots_df['associationCode'] = spots_df['associationCode'].astype('string')
spots_df['summitCode'] = spots_df['summitCode'].astype('string')
spots_df['mode'] = spots_df['mode'].astype('string')
spots_df['frequency'] = spots_df['frequency'].astype('float')

spots_df['timeStamp'] = pd.to_datetime(spots_df['timeStamp'])
# add full summit codes column to spots_df DataFrame
spots_df['summit'] = spots_df['associationCode']+'/'+spots_df['summitCode']

In [68]:
spots_df.dtypes

id                            int64
userID                        int64
timeStamp            datetime64[ns]
comments                     object
callsign                     object
associationCode      string[python]
summitCode           string[python]
activatorCallsign    string[python]
activatorName                object
frequency                   float64
mode                 string[python]
summitDetails                object
highlightColor               object
summit               string[python]
dtype: object

And we can check data for all spots downloaded:

In [69]:
spots_df

id  userID           timeStamp                              comments  \
0   80877   19344 2025-04-14 21:42:02   [RBNHole] at VE6JY 19 WPM 22 dB SNR   
1   80876    1058 2025-04-14 21:34:31      [KJ7QEL{GB}]: cq [SOTA Spotter]    
2   80875   19344 2025-04-14 21:31:04  [RBNHole] at K2PO/7 19 WPM 11 dB SNR   
3   80874   72923 2025-04-14 21:28:15                                         
4   80873   33443 2025-04-14 21:27:59                      [APRS2SOTA] TEST   
5   80872   72923 2025-04-14 21:21:10                                         
6   80871   72923 2025-04-14 21:10:06                 BIG SPACES PSE new op   
7   80870   72923 2025-04-14 21:09:59                 BIG SPACES PSE new op   
8   80869   19344 2025-04-14 21:00:02    [RBNHole] at ND7K 19 WPM 26 dB SNR   
9   80868   49395 2025-04-14 20:58:03                             [sotl.as]   
10  80867   11657 2025-04-14 20:56:07                                  None   
11  80866   19344 2025-04-14 20:56:03  [RBNHole] at KD7EFG 17 WPM 31 dB SNR   
12  80865   19344 2025-04-14 20:55:02   [RBNHole] at KO7SS 25 WPM 33 dB SNR   
13  80864   10513 2025-04-14 20:53:04      +/- QRM. In 5 minutes  [sotl.as]   
14  80863   72923 2025-04-14 20:52:36                 BIG SPACES PSE new op   
15  80862   19344 2025-04-14 20:51:01    [RBNHole] at NT6Q 17 WPM 22 dB SNR   
16  80861   19344 2025-04-14 20:51:01  [RBNHole] at WA7LNW 25 WPM 28 dB SNR   

     callsign associationCode summitCode activatorCallsign activatorName  \
0     RBNHOLE             W7O     WV-144             N7KOM           Tim   
1         SMS             W7O     WV-114           N7KOM/P           Tim   
2     RBNHOLE             W7O     WV-144             N7KOM           Tim   
3     SOTAmat              W6     NS-264              KE6X       Randall   
4   APRS2SOTA              SP     BS-002          SQ9PBR/P         Pawel   
5     SOTAmat              W6     NS-264              KE6X       Randall   
6     SOTAmat              W6     NS-264              KE6X       Randall   
7     SOTAmat              W6     NS-264              KE6X       Randall   
8     RBNHOLE             W7A     YV-001              W8JH           Joe   
9        WV0X             W5N     PW-020              WV0X        Szymon   
10       WU7H              W6     CT-021            KN6DMO          Luis   
11    RBNHOLE              W6     CT-021            KN6DMO          Luis   
12    RBNHOLE             W7A     AW-001             K7GUD       Patrick   
13       W8JH             W7A     YV-001              W8JH           Joe   
14    SOTAmat              W6     NS-264              KE6X       Randall   
15    RBNHOLE              W6     CT-021            KN6DMO          Luis   
16    RBNHOLE             W7A     AW-001             K7GUD       Patrick   

    frequency mode                     summitDetails highlightColor  \
0      7.0620   CW     Lenhart Butte, 645m, 1 points           None   
1      7.0620   CW             4420, 1347m, 2 points           None   
2      7.0620   CW     Lenhart Butte, 645m, 1 points           None   
3     14.2730  SSB        Boca Hill, 2042m, 6 points           None   
4    145.5500   FM      Ostry Wierch, 930m, 6 points           None   
5     14.2300  SSB        Boca Hill, 2042m, 6 points           None   
6      7.1800  SSB        Boca Hill, 2042m, 6 points           None   
7      7.1800  SSB        Boca Hill, 2042m, 6 points           None   
8     14.0620   CW   Tritle, Mount, 2372m, 10 points           None   
9      0.0000                  9420, 2871m, 8 points           None   
10    14.0610   CW  Vetter Mountain, 1801m, 6 points           None   
11    14.0610   CW  Vetter Mountain, 1801m, 6 points           None   
12     7.0617   CW    Mount Lemmon, 2792m, 10 points           None   
13    14.0600   CW   Tritle, Mount, 2372m, 10 points           None   
14     0.0000             Boca Hill, 2042m, 6 points           None   
15     7.0610   CW  Vetter Mountain, 1801m, 6 points           None   
16    10

### Import data from SOTA Database

As you can see, spots do not provide details on activated summits, beside their reference codes. To retrieve this, we need to use SOTA Database. We can use SOTA API for this, but it will be time consuming, so I'll download CSV file with all the summits, stored at https://www.sotadata.org.uk/summitslist.csv. You can also use local file stored in repository - in this case you need to comment line 5 and uncomment line 6 in the cell below.

CSV file includes following columns:
- 0: Summit's reference code,
- 1: association name,
- 2: region's name,
- 3: summit's name,
- 4: altitude (in meters),
- 5: altitude (in feets),
- 6 and 8: longitude,
- 7 and 9: latitude,
- 10: points assigned to a summit in SOTA programme,
- 11: bonus points for activation in winter,
- 12: date when summit has been added to SOTA programme,
- 13: summit's expiry date in SOTA programme,
- 14: number of activations
- 15: last activation date,
- 16: last logged activator.

To make data usable, I need to convert it to proper format. Then, I'll re-index dataframe created so summits are indexed with their reference code.

In [70]:
# create DataFrame based on csv file with all the summits saved (regularly updated
# from https://www.sotadata.org.uk/summitslist.csv, and converting the datatypes
# first row of CSV file is a header, so should be ignored

SOTA_summits_df = pd.read_csv('https://www.sotadata.org.uk/summitslist.csv', skiprows = 1, dtype = {
#SOTA_summits_df = pd.read_csv('summitslist.csv', skiprows = 1, dtype = {
    0: 'string', # summit's reference code
    1: 'string', # association name
    2: 'string', # region name
    3: 'string', # summits name
    4: 'int', # altitude in meters
    5: 'int', # altitude in feets
    6: 'string', # longitude
    7: 'string', # latitude
    8: 'float', # longitude
    9: 'float', # latitude
    10: 'int', # SOTA points for summit
    11: 'int', # bonus points for activations in winter
#    12: not relevant - date when summit was added to SOTA
#    13: not relevant - summit's expiry date
    14: 'int', # number of activiations
#    15: not relevant - last activation date
#    16: not relevant - last activation logged
}
)

SOTA_summits_df = SOTA_summits_df.set_index('SummitCode')

In [71]:
SOTA_summits_df.head()

AssociationName                 RegionName  \
SummitCode                                              
3Y/BV-001    Bouvet Island  Bouvetøya (Bouvet Island)   
4O/IC-001       Montenegro            Istok Crne Gore   
4O/IC-002       Montenegro            Istok Crne Gore   
4O/IC-003       Montenegro            Istok Crne Gore   
4O/IC-004       Montenegro            Istok Crne Gore   

                                  SummitName  AltM  AltFt GridRef1  GridRef2  \
SummitCode                                                                     
3Y/BV-001                         Olavtoppen   780   2559   3.3565  -54.4104   
4O/IC-001                         Maja Rosit  2524   8280  19.8505   42.4795   
4O/IC-002                          Kom kučki  2487   8159  19.6417   42.6807   
4O/IC-003   Veliki vrh (Maja Gurt e Zjarmit)  2480   8136  19.7872   42.4971   
4O/IC-004                    Kom vasojevićki  2461   8074  19.6677   42.6879   

            Longitude  Latitude  Points  BonusPoints   ValidFrom     ValidTo  \
SummitCode                                                                     
3Y/BV-001      3.3565  -54.4104      10            3  01/03/2018  31/12/2099   
4O/IC-001     19.8505   42.4795      10            3  01/03/2019  31/12/2099   
4O/IC-002     19.6417   42.6807      10            3  01/03/2019  31/12/2099   
4O/IC-003     19.7872   42.4971      10            3  01/03/2019  31/12/2099   
4O/IC-004     19.6677   42.6879      10            3  01/03/2019  31/12/2099   

            ActivationCount ActivationDate ActivationCall  
SummitCode                                                 
3Y/BV-001                 0            NaN            NaN  
4O/IC-001                 1     27/07/2022    4O/SQ9MDF/P  
4O/IC-002                 0            NaN            NaN  
4O/IC-003                 0            NaN            NaN  
4O/IC-004                 0            NaN            NaN

In [72]:
SOTA_summits_df.dtypes

AssociationName    string[python]
RegionName         string[python]
SummitName         string[python]
AltM                        int64
AltFt                       int64
GridRef1           string[python]
GridRef2           string[python]
Longitude                 float64
Latitude                  float64
Points                      int64
BonusPoints                 int64
ValidFrom                  object
ValidTo                    object
ActivationCount             int64
ActivationDate             object
ActivationCall             object
dtype: object

### Spots analysis

If you look at the spots_dfdataframe above, it may includes duplicates. The same activation may be spotted more than once. For visualisation purpose, I'll take the latest spot for operator-summit pair. The logic behind this is:
- it's highly unlikely to activate two summits by the same person within an hour,
- if two spots for the same operator-summit pair are exactly the same, the newest one is a confirmation "I'm still activating", so I'll take it to countdown time,
- if operator changed mode or band, the newest spot present the latest available information about the activation.

In [73]:
# drop duplicated activator-summit pairs from spots_df to avoid double visualisation for them
# only last spot sent by activator on a summit is considered

spots_df = spots_df.drop_duplicates(subset = ['activatorCallsign', 'summit'])
print(f'{len(spots_df)} found without duplicates.')

8 found without duplicates.


Now I'm going to prepare spots_df dataframe for visualisation by adding colums, where data from other dataframes (bands_df, modes_df and summits_df) will be copied. Temporary they will be filled with ```None``` values.
When it's done, spots_df dataframe will be reindexed to make sure indexes are consecutive numbers, therefore we'll avoid errors when looping throught its rows

In [74]:
# create empty columns for data required for visualisation (will b filled with database data)
# then re-index this dataframe

spots_df['longitude'] = None
spots_df['latitude'] = None
spots_df['points'] = None
spots_df['summitName'] = None
spots_df['mode_color'] = None
spots_df['band'] = None
spots_df['band_color'] = None
spots_df = spots_df.reset_index(drop = True)

/var/folders/yp/j6434ggs29b1w_s9sl20cn2w0000gn/T/ipykernel_26728/3895443260.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  spots_df['longitude'] = None
/var/folders/yp/j6434ggs29b1w_s9sl20cn2w0000gn/T/ipykernel_26728/3895443260.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  spots_df['latitude'] = None
/var/folders/yp/j6434ggs29b1w_s9sl20cn2w0000gn/T/ipykernel_26728/3895443260.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexe

The truth is that spots sometimes are sent with errors. For this analysis the most important is when summit code spotter is not included in SOTA Database. It may be a result of three situations:
- spotter's typo,
- spotter is not sure of summits references, but wants to forward the frequency to chasers,
- spotted summit has been added to SOTA database after last refresh of CSV files used.

To track if this happen in current list of spots, I'm creating a list to store errors found in next step.
Create a list to handle errors - misspelled codes or updated database.

In [75]:
# list to keep summit codes not found in SOTA Database file
# List of summits is periodically updated, but typos in summits codes in spots are also common
summits_errors = []

Now we are ready to match 
With a loop through all spots (hope you remember we've removed duplicates) script will check if summit reference is present in SOTA Databased. If no, summit will be added to errors list with no further action.

If summit is found, following data will be copied from SOTA database to spots_df DataFrame for each spot:
- longitude and latutide,
- number of SOTA points (from 1 to 10),
- name of the summit.

Also, based on spots data, time since spot's update (in hours), activation's mode and band are populated. For mode and band, declared color will be also assigned.

For wisualisation purpose, I'm also creating popup Series to store activation description to be displayed when marker is clicked on the map. Sample looks like this:

> Summit Red Mountain - W6/CT-133 (4 points) activated by KX6I on 14.061 - CW 3 minutes ago.


In [76]:
# copying relevant data for visualisation from SOTA database extract to spots dataframe
# also adding time since spot in hour fraction and description of spot
# adding previously defined colorcodes for band and mode to each spot, together with a band for each one
# popup column provides a summary of activation to be displayed on map
for i in range(0, len(spots_df)):
    # if summits data are correct,prepare spot's data for visualisation
    if spots_df.loc[i, ('summit')].upper() in SOTA_summits_df.index:
        spots_df.loc[i, ('longitude')] = SOTA_summits_df['Longitude'][spots_df.loc[i, 'summit']]
        spots_df.loc[i, ('latitude')] = SOTA_summits_df['Latitude'][spots_df.loc[i, 'summit']]
        spots_df.loc[i, ('points')] = SOTA_summits_df['Points'][spots_df.loc[i, 'summit']]
        spots_df.loc[i, ('summitName')] = SOTA_summits_df['SummitName'][spots_df.loc[i, 'summit']]
        spots_df.loc[i, ('time_since_spot')] = datetime.utcnow()-spots_df.loc[i, ('timeStamp')]
        spots_df.loc[i, ('time_since_spot')] = spots_df.loc[i, ('time_since_spot')]/timedelta(hours=1)
        spots_df.loc[i, ('popup')] = f"Summit {spots_df.loc[i, ('summitName')].title()} - {spots_df.loc[i, ('summit')]} ({spots_df.loc[i, ('points')]} points)\nactivated by {spots_df.loc[i, ('activatorCallsign')].upper()}\non {spots_df.loc[i, ('frequency')]} - {spots_df.loc[i, ('mode')].upper()}\n{round(spots_df.loc[i, ('time_since_spot')]*60)} minutes ago\n."
        spots_df.loc[i, ('mode')] = spots_df.loc[i, ('mode')].upper() # for correct match between spots and summits DataFrames
        for band in bands_df.index: # assess band based on frequency spotted
            if (spots_df.loc[i, ('frequency')] >= bands_df['lower_freq'][band]) and (spots_df.loc[i, ('frequency')] <= bands_df['upper_freq'][band]):
                spots_df.loc[i, ('band_color')] = bands_df['color'][band]
                spots_df.loc[i, ('band')] = band
        for j in modes_df.index:
            if spots_df.loc[i, ('mode')] == modes_df.iloc[j]['mode'].upper():
                spots_df.loc[i,('mode_color')] = modes_df.iloc[j]['color']
    # if summit isn't found in database, print warning, save it on a list and leave their data with None
    else:
        print(f"Summit {spots_df.loc[i, ('summit')]} activated by {spots_df.loc[i, ('activatorCallsign')].upper()} on {spots_df.loc[i, ('frequency')]} - {spots_df.loc[i, ('mode')].upper()}  NOT FOUND.")
        summits_errors.append({spots_df.loc[i, ('summit')]})

/var/folders/yp/j6434ggs29b1w_s9sl20cn2w0000gn/T/ipykernel_26728/3657109216.py:13: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '0.137939715' has dtype incompatible with timedelta64[ns], please explicitly cast to a compatible dtype first.
  spots_df.loc[i, ('time_since_spot')] = spots_df.loc[i, ('time_since_spot')]/timedelta(hours=1)


The spots_df DataFrame looks like this and is ready for visualisation.

In [77]:
spots_df.head()

id  userID           timeStamp                             comments  \
0  80877   19344 2025-04-14 21:42:02  [RBNHole] at VE6JY 19 WPM 22 dB SNR   
1  80876    1058 2025-04-14 21:34:31     [KJ7QEL{GB}]: cq [SOTA Spotter]    
2  80874   72923 2025-04-14 21:28:15                                        
3  80873   33443 2025-04-14 21:27:59                     [APRS2SOTA] TEST   
4  80869   19344 2025-04-14 21:00:02   [RBNHole] at ND7K 19 WPM 26 dB SNR   

    callsign associationCode summitCode activatorCallsign activatorName  \
0    RBNHOLE             W7O     WV-144             N7KOM           Tim   
1        SMS             W7O     WV-114           N7KOM/P           Tim   
2    SOTAmat              W6     NS-264              KE6X       Randall   
3  APRS2SOTA              SP     BS-002          SQ9PBR/P         Pawel   
4    RBNHOLE             W7A     YV-001              W8JH           Joe   

   frequency  ...      summit longitude latitude points     summitName  \
0      7.062  ...  W7O/WV-144  -122.167  45.3502      1  Lenhart Butte   
1      7.062  ...  W7O/WV-114  -122.442  44.6223      2           4420   
2     14.273  ...   W6/NS-264 -120.1196  39.3812      6      Boca Hill   
3    145.550  ...   SP/BS-002    21.123  49.4385      6   Ostry Wierch   
4     14.062  ...  W7A/YV-001 -112.4432  34.4169     10  Tritle, Mount   

  mode_color     band band_color time_since_spot  \
0        red    7 MHz        red         0.13794   
1        red    7 MHz        red        0.263223   
2       blue   14 MHz     orange        0.367669   
3     yellow  144 MHz       blue        0.372114   
4        red   14 MHz     orange        0.837949   

                                               popup  
0  Summit Lenhart Butte - W7O/WV-144 (1 points)\n...  
1  Summit 4420 - W7O/WV-114 (2 points)\nactivated...  
2  Summit Boca Hill - W6/NS-264 (6 points)\nactiv...  
3  Summit Ostry Wierch - SP/BS-002 (6 points)\nac...  
4  Summit Tritle, Mount - W7A/YV-001 (10 points)\...  

[5 rows x 23 columns]

But before we move forward, I'll save incorrectly spotted rows in summits_errors.txt file. I'm using it to check if CSV file with summits database stored locally needs an update.

In [78]:
# save errors to file
if len(summits_errors) != 0:
    with open('summits_errors.txt', 'a') as f:
        for error in summits_errors:
            f.write(f'{error}\n')

## Visualisation of spots

In this script, spots will be visualised on the map with the use of Folium module. First, we'll create a map centered near Kraków (the city where I live) with the shaded terrain background.

In [80]:
# create a map in Folium - just works
activations_map = folium.Map(location = [50,20], # map is centered on Kraków - city where I live
                        tiles="OpenStreetMap", 
                        zoom_start=2 # show whole world at once
                            )

Now every spot kept in spots_df DataFrame will be visualised as a circle on the map with the use of CirleMarker markers. Each point provides following information:
- activated summit's **location** - circle's center;
- **time** since spot - circle's radius;
- activated **band** - circle's filling;
- **mode** use for activiation - circle's border.

Also, previously defined pop-up text is available when user clicks activation marker.

In [81]:
# add spots to a map
for i in range(0, len(spots_df)): # add point for every spot (with duplicates removed)
    if spots_df.loc[i, ('longitude')] != None:  # ignore spots where no reference data in SOTA database was found
        folium.CircleMarker(
        location = [spots_df.loc[i, ('latitude')], spots_df.loc[i, ('longitude')]], # spot's location)
        radius = (1 - spots_df.loc[i, ('time_since_spot')])*15, # radius is proportional to time from sending
                                                                # the spot. The newest spot, the larger circle
        popup = spots_df.loc[i, ("popup")],
        fill_color = spots_df.loc[i, ('band_color')], # circle's fill represents activation's band
        weight = 3,
        color = spots_df.loc[i, ('mode_color')], # border color represents activation's mode
        fill_opacity = 1
    ).add_to(activations_map)
    
display(activations_map)

Enjoy tracking current activations - hope this script will support you in chasing.